lua/customroles/soulbound.lua

```diff
@@ -0,0 +1,673 @@
+ROLE.blockshopconvars = true
+
+RegisterRole(ROLE)
+
+hook.Add("TTTRoleSpawnsArtificially", "Soulbound_TTTRoleSpawnsArtificially", function(role)
+    if role == ROLE_SOULBOUND and util.CanRoleSpawn(ROLE_SOULMAGE) then
+        return true
+    end
+end)
+
+--------------------------
+-- ABILITY REGISTRATION --
+--------------------------
+
+SOULBOUND = {
+    Abilities = {}
+}
+
+function SOULBOUND:RegisterAbility(ability)
+    ability.Id = ability.Id or ability.id or ability.ID
+
+    if SOULBOUND.Abilities[ability.Id] then
+        ErrorNoHalt("[SOULBOUND] Soulbound ability already exists with ID '" .. ability.Id .. "'\n")
+        return
+    end
+
+    local enabled = CreateConVar("ttt_soulbound_" .. ability.Id .. "_enabled", "1", FCVAR_REPLICATED)
+    ability.Enabled = function()
+        return enabled:GetBool()
+    end
+    table.insert(ROLE_CONVARS[ROLE_SOULBOUND], {
+        cvar = "ttt_soulbound_" .. ability.Id .. "_enabled",
+        type = ROLE_CONVAR_TYPE_BOOL
+    })
+
+    SOULBOUND.Abilities[ability.Id] = ability
+end
+
+local abilityFiles, _ = file.Find("soulbound_abilities/*.lua", "LUA")
+for _, fil in ipairs(abilityFiles) do
+    if SERVER then AddCSLuaFile("soulbound_abilities/" .. fil) end
+    include("soulbound_abilities/" .. fil)
+end
+
+if SERVER then
+    util.AddNetworkString("TTT_SoulboundBuyAbility")
+    util.AddNetworkString("TTT_SoulboundUseAbility")
+
+    ---------------
+    -- ABILITIES --
+    ---------------
+
+    net.Receive("TTT_SoulboundUseAbility", function(len, ply)
+        local num = net.ReadUInt(4)
+        if not ply:IsSoulbound() then return end
+
+        local id = ply:GetNWString("TTTSoulboundAbility" .. tostring(num), "")
+        if #id == 0 then return end
+
+        local ability = SOULBOUND.Abilities[id]
+        if not ability:Enabled() then return end
+
+        local target = ply:GetObserverMode() ~= OBS_MODE_ROAMING and ply:GetObserverTarget() or nil
+        if not ability:Condition(ply, target) then return end
+        if not ability.Use then return end
+        ability:Use(ply, target)
+    end)
+
+    -----------------------
+    -- PASSIVE ABILITIES --
+    -----------------------
+
+    hook.Add("Think", "Soulbound_Think", function()
+        for _, p in ipairs(player.GetAll()) do
+            if p:IsSoulbound() then
+                local max = soulbound_max_abilities:GetInt()
+                for i = 1, max do
+                    local id = p:GetNWString("TTTSoulboundAbility" .. tostring(i), "")
+                    if #id == 0 then break end
+
+                    local ability = SOULBOUND.Abilities[id]
+                    if not ability:Enabled() then return end
+
+                    local target = p:GetObserverMode() ~= OBS_MODE_ROAMING and p:GetObserverTarget() or nil
+                    if not ability:Condition(p, target) then return end
+                    if not ability.Passive then continue end
```

Comment:
Should probably move this check up to right after the ability is fetched to short-circuit any logic in `Enabled` and `Condition`